AdblockPlusSafari/RootController.m

```diff
@@ -48,4 +72,94 @@ - (void)setViewControllers:(NSArray<UIViewController *> *)viewControllers animat
   [super setViewControllers:viewControllers animated:animated];
 }
 
+- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSString *,id> *)change context:(void *)context
+{
+  if ([keyPath isEqualToString:NSStringFromSelector(@selector(activated))]
+      || [keyPath isEqualToString:NSStringFromSelector(@selector(needsDisplayErrorDialog))]) {
+    if (_readyToShowDialog) {
+      [self showDialogIfNeeded];
+    }
+  } else {
+    [super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
+  }
+}
+
+- (void)setAdblockPlus:(AdblockPlusExtras *)adblockPlus
+{
+  NSArray<NSString *> *keyPaths = @[NSStringFromSelector(@selector(activated)),
+                                    NSStringFromSelector(@selector(needsDisplayErrorDialog))];
+
+  for (NSString *keyPath in keyPaths) {
+    [_adblockPlus removeObserver:self
+                      forKeyPath:keyPath];
+  }
+  _adblockPlus = adblockPlus;
+  for (NSString *keyPath in keyPaths) {
+    [_adblockPlus addObserver:self
+                   forKeyPath:keyPath
+                      options:NSKeyValueObservingOptionInitial|NSKeyValueObservingOptionNew
+                      context:nil];
+  }
+}
+
+#pragma mark - Private
+
+- (void)showDialogIfNeeded
+{
```

Comment:
RootController is always present. Welcome dialog is always presented before error dialog. Both dialogs are not presented at once.